autonat/autonat-v2.md

```diff
@@ -0,0 +1,137 @@
+well-behaved in the network: A node that's behind a doesn't need to
+advertise its (undialable) addresses to the rest of the network, preventing
+superfluous dials from other peers. Furthermore, it might actively seek to
+improve its connectivity by finding a relay server, which would allow other
+peers to establish a relayed connection.
+
+`autonat v1` protocol was designed for a node to determine whether the node
+was publically rechable. However, it could not be tweaked to obtain more 
+granular information about which of the nodes addresses were publically 
+reachable. `autonat v2` allows nodes to determine reachability for individual
+addresses. Using `autonat v2` nodes can build an address pipeline where they
+can test individual addresses discovered by different sources like identify, 
+upnp mappings, circuit addresses etc for reachability.
+
+
+## AutoNAT V2 Protocol
+
+The AutoNAT V2 Protocol uses the Protocol ID `/libp2p/autonat/2.0.0`. The node
+wishing to determine reachability of a given address sends a `DialRequest` 
+message. The `DialRequest` message contains a single multiaddress and a nonce 
+of 8 random bytes. Upon receiving  this message, the peer dials this address 
+and sends the nonce in the `Dial` message. The peer MUST NOT dial any address 
+other than the address in the DialRequest. The initiator SHOULD check that the 
+nonce received in the `Dial` message is the same as the nonce the initiator 
+sent in the `DialRequest` message. If the nonce received in the `Dial` message 
+is different the initiator should discard this `DialResponse`
+
+If the dial fails, the receiver sends a `DialResponse` message with the
+`ResponseStatus` `E_DIAL_ERROR`. If the dial completes successfully, it 
+sends a `DialResponse` with the `ResponseStatus` `OK`. Nodes MUST NOT send a 
+`E_DIAL_ERROR` status when they didn't attempt a dial. 
+
+In order to prevent attacks like the one described in [RFC 3489, Section
+12.1.1](https://www.rfc-editor.org/rfc/rfc3489#section-12.1.1) (see excerpt
+below), implementations MUST NOT dial any multiaddress unless it is based on the
+IP address the requesting node is observed as. This restriction as well implies 
+that implementations MUST NOT accept dial requests via relayed connections as 
+one can not validate the IP address of the requesting node.
+
+> RFC 3489 12.1.1 Attack I: DDOS Against a Target
+>
+> In this case, the attacker provides a large number of clients with the same
+> faked MAPPED-ADDRESS that points to the intended target. This will trick all
+> the STUN clients into thinking that their addresses are equal to that of the
+> target. The clients then hand out that address in order to receive traffic on
+> it (for example, in SIP or H.323 messages). However, all of that traffic
+> becomes focused at the intended target. The attack can provide substantial
+> amplification, especially when used with clients that are using STUN to enable
+> multimedia applications.
```

Comment:
Sounds like after the discussion in https://github.com/libp2p/specs/issues/536#issuecomment-1497860644, we might not need this after all? But let's hear what @mxinden has to say first.